noir_stdlib/src/meta/expr.nr

```diff
@@ -110,4 +110,305 @@ impl Expr {
     // docs:start:is_continue
     fn is_continue(self) -> bool {}
     // docs:end:is_continue
+
+    // docs:start:map
+    fn map<Env>(self, f: fn[Env](Expr) -> Option<Expr>) -> Expr {
+        // docs:end:map
+        let result = f(self);
+        let result = result.or_else(|| map_array(self, f));
+        let result = result.or_else(|| map_assign(self, f));
```

Comment:
I thought of an issue with our current design of this function.
The use case is to provide an easy way for users to traverse the Ast and modify the nodes they want to. For example, adding an argument to each function call.

However, since we don't apply the function recursively after we already return `Some` on a node, we can run into an issue where `foo(bar(baz))` would turn into `foo(ctx, bar(baz))` instead of `foo(ctx, bar(ctx, baz))`. I think the easiest way to fix this issue is to apply `f` _after_ we recur on the subexpressions instead of before - and to apply it unconditionally.

E.g.
```rs
let result = map_array(self, f);
let result = result.or_else(|| map_assign(self, f));
...
let result = result.or_else(|| map_unsafe(self, f));

if result.is_some() {
    let modified = f(result.unwrap_unchecked());
    if modified.is_some() {
        modified
    } else {
        result
    }
} else {
    self
}
// could also be:
// result.and_then(|expr| f(expr).unwrap_or(expr)).unwrap_or(self)
```

Reply:
> we can run into an issue where foo(bar(baz)) would turn into foo(ctx, bar(baz)) instead of foo(ctx, bar(ctx, baz))

Right, I noticed that too. In the "inject" example what I ended up doing is, if `Expr` is a function call, add the `context` argument but also call `mutate` on each of the arguments to continue injecting `context`. I think doing it that way is a bit more cumbersome but it also gives some more flexibility. For example if I only wanted to add a `context` argument to top-level calls but not calls nested in those, I couldn't do it if sub-expressions were always mutated (or maybe I could if I kept some external state to know where I am, but maybe it would be more confusing).

Let me know if you'd still prefer to always recur.

Reply:
Actually, I'll change it to always recur. We can always re-evaluate this as we implement more use cases. Maybe we never bump into one that doesn't need to always recur.